### DISCUSSION TITLE
Best Practice for monitoring the relays

### DISCUSSION THREAD

**Discussion by throny** [nusenu/ansible-relayor#227] [General]
First of all: thanks for the good introduction on medium which led me here and made me set up a cluster of 4 relays.
Now I saw that its possible to gather metrics with Prometheus and I'd really like to do so.

However it's not really straight forward, it would be great to have a sample on how to setup Prometheus and gather i.e. bandwith information.
That would help to give people like me a better entry to the whole project.

**Comment by nusenu** [comment]
The prometheus exporter in tor is new and unfortunately still very limited. It does not provide a lot of datapoints. For example you can not do bandwidth graphs using tor's native exporter.
For reference here is the upstream ticket that is open since over a year (2020-11-16) without any real progress: https://gitlab.torproject.org/tpo/core/tor/-/issues/40194

As for the "how to use the new prometheus support in relayor": I'll create a wiki page for that. It should be there when the first non "rc" release with prometheus support is published.
I might also add another feature to generate a blackbox exporter tcp check for all ORPorts.

**Comment by throny** [comment]
Hi again,

it seems that there happened a lot to the prometheus integration. However, I'm still struggling with setting that up and I couldn't find a guide in Wiki.
If I understood correctly, the prometheus server has to be in the inventory too? Can't I just set up a prometheus somewhere (where no tor instance is running) and let it be collected there?
Sorry for these stupid questions, I'm a little lost here.

Edit: Nevermind, just saw that you posted this in the discussions. Will have a look!

**Comment by nusenu** [comment]
Hi,

> If I understood correctly, the prometheus server has to be in the
> inventory too? Can't I just set up a prometheus somewhere (where no
> tor instance is running) and let it be collected there?

Yes, you can use a prometheus server that is not running tor itself,
but it has to be in your ansible inventory if you want to use
the automation. The does not install tor on all your hosts in your inventory.

Did you see the HowTo?
https://github.com/nusenu/ansible-relayor/discussions/239

**Comment by throny** [comment]
Really appreciate the work you put in here! 
I will have to take some more time to learn how to safely run Prometheus in the first place, it seems.
